pandas/_libs/groupby.pyx

```diff
@@ -1794,6 +1794,143 @@ cdef group_min_max(
         )
 
 
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def group_idxmin_idxmax(
+    intp_t[:, ::1] out,
+    int64_t[::1] counts,
+    ndarray[numeric_object_t, ndim=2] values,
+    const intp_t[::1] labels,
+    Py_ssize_t min_count=-1,
+    bint is_datetimelike=False,
+    const uint8_t[:, ::1] mask=None,
+    str name="idxmin",
+    bint skipna=True,
+    uint8_t[:, ::1] result_mask=None,
+):
+    """
+    Compute index of minimum/maximum of columns of `values`, in row groups `labels`.
+
+    This function only computes the row number where the minimum/maximum occurs, we'll
+    take the corresponding index value after this function.
+
+    Parameters
+    ----------
+    out : np.ndarray[int64_t, ndim=2]
+        Array to store result in.
+    counts : np.ndarray[int64]
+        Input as a zeroed array, populated by group sizes during algorithm
+    values : np.ndarray[numeric_object_t, ndim=2]
+        Values to find column-wise min/max of.
+    labels : np.ndarray[np.intp]
+        Labels to group by.
+    min_count : Py_ssize_t, default -1
+        The minimum number of non-NA group elements, NA result if threshold
+        is not met.
+    is_datetimelike : bool
+        True if `values` contains datetime-like entries.
+    name : {"idxmin", "idxmax"}, default "idxmin"
+        Whether to compute idxmin or idxmax.
+    mask : ndarray[bool, ndim=2], optional
+        If not None, indices represent missing values,
+        otherwise the mask will not be used
+    skipna : bool, default True
+        Flag to ignore nan values during truth testing
+    result_mask : ndarray[bool, ndim=2], optional
+        If not None, these specify locations in the output that are NA.
+        Modified in-place.
+
+    Notes
+    -----
+    This method modifies the `out` parameter, rather than returning an object.
+    `counts` is modified to hold group sizes
+    """
+    cdef:
+        Py_ssize_t i, j, N, K, lab
+        numeric_object_t val
+        numeric_object_t[:, ::1] group_min_or_max
+        bint uses_mask = mask is not None
+        bint isna_entry
+        bint compute_max = name == "idxmax"
+
+    assert name == "idxmin" or name == "idxmax"
+
+    # TODO(cython3):
+    # Instead of `labels.shape[0]` use `len(labels)`
+    if not len(values) == labels.shape[0]:
+        raise AssertionError("len(index) != len(labels)")
+
+    group_min_or_max = np.empty_like(out, dtype=values.dtype)
+    # Using -2 to indicate a label is never seen
```

Comment:
Is the use of this sentinel custom to this function or do we have a prior art for it in the code base?

Reply:
I believe this is novel.

Reply:
I have a slight preference to just create an `unobserved_mask` array. My main apprehension with using a sentinel like this is that it requires the caller / callee to both maintain that knowledge, whereas a mask is much more explicit

Reply:
At first glance, it seems like that would require a large refactor of WrappedCythonOp. Any recommendations for how this new mask could be used in the existing pattern?

Reply:
This can't be passed through **kwargs?

Reply:
I see - so we create a NumPy array and pass via kwargs to the Cython function. I also do not like using `-2` as a sentinel, and so definitely wanted to try this out. 

However I also don't like the pattern of passing values via kwargs to be mutated and using a mask requires an extra assignment each loop to update it. For the following code, I'm seeing 76.6ms vs 80.0ms, so about a 4.4% slowdown.

<details>
<summary>Code</summary>

```
size = 10_000_000

df = pd.DataFrame(
    {
        "a": np.random.randint(0, 100, size),
        "b": np.random.randint(0, 1000, size),
    }
)

%timeit df.groupby("a").idxmin()
```

</details>

If it was a clear win from a design standpoint I'd be more for it, but to me it's more of a wash. The minor perf hit makes me lean toward current state (-2 sentinel), but not strongly so. Thoughts?

Reply:
The performance here doesn't matter; it is a clear design win because the pattern is reusable. When encoding a sentinel you would either want an enumeration to map codes to some type of value or you are asking for tribal knowledge of caller/callee behavior. The latter is undesirable.

Reply:
Another consideration point is that if we wanted to start using bitmasks like https://github.com/pandas-dev/pandas/pull/54506 passing in an extra Boolean buffer-compliant argument would require no change, whereas the sentinel solution would not work with that. 

Note that the PR linked currently exposes only has a read only buffer through the buffer protocol, but should be extensible to read write. At a minimum, a BitmaskArray object itself supports setitem

Reply:
At the sprint Richard and I discussed the -2 thing and I mentioned having another use case in mind where I was thinking of something similar.  But the usage I was thinking of would mean a _different_ sentinel than it does here.

Will makes a fair point about tribal knowledge.  Spitballing: we _could_ make an enum for both sentinels, establishing precedent for using more of them in the future as use cases arise.  If an enum isn't too costly, then I think I'd be more comfortable with that than with hard-coding -2s in either use case.  And in principle we could make -1 part of the enum to de-tribalize the knowledge of how we use -1.

Reply:
When I tested it out, I only tested a single column. When using multiple columns, we make multiple calls to WrappedCythonOp. An issue arises because of the following two facts:

 - We mutate the mask in the Cython code.
 - When used with object dtype, we can't pre-populate `group_min_or_max` with min/max values because we don't know what objects these are.

As a result, I only see two options for using the mask approach:

 - reset the mask each time we iterate over the columns; or
 - use a -2-like sentinel in the Cython code alongside the mask

Reply:
@WillAyd - if we're still going the mask route, I think we'd prefer the first option above? I can put up a commit showing what that'd look like.

Reply:
Is a 2D mask not an option? Didn't look very closely but I thought we e already done that in other parts of the code base?

Reply:
Ah - I mispoke (as looking at the case where each column is a different data types). We indeed do operate by blocks, but we still need to loop over blocks.

Reply:
> When used with object dtype, we can't pre-populate group_min_or_max with min/max values because we don't know what objects these are. This means we need a "hasn't been seen" indicator to use prior to testing "is the current min/max great/less than the current value".

`_libs.algos` has Infinity and NegInfinity that seem similar to what you're describing.  See get_rank_nan_fill_val.

Reply:
Thanks @jbrockmendel! I didn't dive into `_get_min_or_max` and assumed it was only non-object dtype. This makes the masked version 72ms (was 80ms before).